src/main/java/de/tum/in/www1/artemis/service/MailService.java

```diff
@@ -153,6 +153,48 @@ public void sendSAML2SetPasswordMail(User user) {
 
     // notification related
 
+    /**
+     * Sets the correct context for the case that the notificationSubject is an Exercise
+     * @param notificationSubject which is an Exercise
+     * @param context that is modified based on the exercise type
+     */
+    private void setExerciseContext(Object notificationSubject, Context context) {
+        if (notificationSubject instanceof QuizExercise) {
+            context.setVariable(EXERCISE_TYPE, "quiz");
+        }
+        else if (notificationSubject instanceof ModelingExercise) {
+            context.setVariable(EXERCISE_TYPE, "modeling");
+        }
+        else if (notificationSubject instanceof TextExercise) {
+            context.setVariable(EXERCISE_TYPE, "text");
+        }
+        else if (notificationSubject instanceof ProgrammingExercise) {
+            context.setVariable(EXERCISE_TYPE, "programming");
+        }
+        else if (notificationSubject instanceof FileUploadExercise) {
+            context.setVariable(EXERCISE_TYPE, "upload");
+        }
+    }
+
+    /**
+     * Sets the context and subject for the case that the notificationSubject is a Post
+     * @param context that is modified
+     * @param notificationSubject which has to be a Post
+     * @param locale used for translations
+     * @return the modified subject of the email
+     */
+    private String setPostContextAndSubject(Context context, Object notificationSubject, Locale locale) {
+        // posts use a different mechanism for the url
+        context.setVariable(NOTIFICATION_URL, NotificationTarget.extractNotificationUrl((Post) notificationSubject, artemisServerUrl.toString()));
+
+        if (locale.toString().equals("en")) {
+            return "New announcement \"" + ((Post) notificationSubject).getTitle() + "\" in course \"" + ((Post) notificationSubject).getCourse().getTitle() + "\"";
+        }
+        else {
+            return "Neue Ankündigung \"" + ((Post) notificationSubject).getTitle() + "\" im Kurs \"" + ((Post) notificationSubject).getCourse().getTitle() + "\"";
+        }
```

Comment:
[nit] I assume there is no translation-like service on the server, as we have on the client, right?
I think we could use formatted strings instead, in order to make it more readable.
Something like:
```
final String newAnnouncementEN = "New announcement \"%s\" in course \"%s\"".
final String newAnnouncementDE = "...";

...
private String setPostContextAndSubject(Context context, Object notificationSubject, Locale locale) {
  ...
  String newAnnouncement = locale.toString().equals("en") ? newAnnouncementEN : newAnnouncementDE ;
  return String.format(newAnnoucement, 
                       ((Post) notificationSubject).getTitle(), 
                       ((Post) notificationSubject).getCourse().getTitle());
}
```
I guess this would also be more easier to port to the translation-like service in the future.

Reply:
This is quite interesting, I have not yet seen this in Java (it reminds me of C). Thank you very much for this idea.
Because this is an issue of code quality and not related to Lorena's work (explicitly) I will implement this (and your other remark) in a small follow-up PR. The logic works and I think this should not hinder this Metis PR from being merged.

Reply:
This is the follow up PR which I will open once this PR has been merged. https://github.com/ls1intum/Artemis/pull/4345